libs/sdk-ui-pivot/src/impl/structure/tableDescriptorTypes.ts

```diff
@@ -0,0 +1,275 @@
+
+/**
+ * Descriptors of all table headers. The table headers are divided into two groups:
+ *
+ * -  Row-column headers; these describe the columns that contain description of each row of data. They are
+ *    constructed from data slicing attributes.
+ *
+ * -  Column headers; these describe the columns that contain the actual data points. They are constructed from
+ *    the data series identified in the result. A data series represents a stream of data points computed for particular
+ *    measure. The data points can be sliced by some attributes and/or scoped by some attributes. Read more about
+ *    that here: {@link IDataSeries}.
+ *
+ *    When the data series use scoping, the table headers should be nicely grouped by the scoping attribute
+ *    element values and form a tree hierarchy. That is why the column headers are further listed as:
+ *
+ *    -  Root Columns = column descriptors representing top-most group
+ *    -  Leaf Columns = column descriptors representing the actual, fully scoped columns that contain the actual
+ *       data points
+ *
+ *    Note: if the data series are not scoped, then no grouping is needed and the root columns and leaf columns will
+ *    contain the same values.
+ *
+ */
+export type TableCols = {
+    /**
+     * All table row headers - these are derived from slicing attributes
+     */
+    readonly sliceCols: SliceCol[];
+
+    /**
+     * Root table headers.
+     */
+    readonly rootDataCols: DataCol[];
+
+    /**
+     * Bottom-most
+     */
+    readonly leafDataCols: Array<DataColGroup | DataColLeaf>;
+
+    /**
+     * Mapping between header ID to header descriptor (useful for lookups as ColDefs 'field' references this)
+     */
+    readonly idToDescriptor: Record<string, AnyCol>;
+
+    /**
+     * All attributes used for grouping table columns.
+     */
+    readonly groupingAttributes: IAttributeDescriptor[];
+};
+
+/**
+ * Ag-Grid colDefs constructed to match the table header descriptors. See {@link TableCols} for more
+ * information about nomenclature.
+ */
+export type TableColDefs = {
+    readonly sliceColDefs: ReadonlyArray<ColDef>;
+    readonly rootDataColDefs: ReadonlyArray<ColDef | ColGroupDef>;
+    readonly leafDataColDefs: ReadonlyArray<ColDef>;
+    readonly idToColDef: Record<string, ColDef | ColGroupDef>;
+};
+
+function isColDef(obj: unknown): obj is ColDef {
+    return (obj as ColDef)?.colId !== undefined;
+}
+
+function isColumn(obj: unknown): obj is Column {
+    return (obj as Column)?.getColDef !== undefined;
+}
+
+/**
+ * Convenience accessor of ag-grid column identifier. This can take either string itself or the
+ * different type of col flying around ag-grid API (ColDef, ColGroupDef or Column).
+ *
+ * For purposes of our table, column ID for ag-grid groups is the groupId.
+ *
+ * @param colOrId - column id (returned as is) or one of the ag-grid's column types
+ */
+export function agColId(colOrId: string | ColDef | ColGroupDef | Column): string {
```

Comment:
I would move this accessor to the agGridUtils.ts where already `isColumn` predicate exists and where is imho better place for this

Reply:
exactly my plan for a followup PR. want to to a bit more shaking there and possibly also rename some of the files.